### PR TITLE
Added the ability to pass lists instead of directories and filters to kerneldbgen

### LINKED ISSUE #3390: Kernel database generation for complex file naming formats
## Description  

In the kerneldbgen application, after name filtering and time filtering, kernel priority is based purely on the alphabetical ordering of the files. In some cases, more complex logic is required to get the best quality kernels ordered first. ISIS should have a way to use more complex logic when ordering the kernels after name and time filtering.

### Simple Example
Let's assume we have one kernel get split in two in a later version. So, in the following list the file naming convention is mis_<start time>_<stop time>_v<version>.bsp and we want to always use the most recent version, so version 153 over version 149.

- mis_190401_190501_v152.bsp (2019-04-01 to 2019-05-01)
- mis_190401_190601_v149.bsp (2019-04-01 to 2019-06-01)
- mis_190501_190601_v152.bsp (2019-05-01 to 2019-06-01)

kerneldbgen can match this set of kernels with the filter "mis_??????_??????_v???.bsp" and it will order them as above (alphabetically). Priority for the kernels is First in Last Out, so here's the priority for each time period:

#### 2019-04-01 to 2019-05-01

1. mis_190401_190601_v149.bsp
1. mis_190401_190501_v152.bsp

#### 2019-05-01 to 2019-01-01

1. mis_190501_190601_v152.bsp
1. mis_190401_190601_v149.bsp

The priority for 2019-04-01 to 2019-05-01 uses the old version over the new version and should be:

1. mis_190401_190501_v152.bsp
1. mis_190401_190601_v149.bsp

### Real Example

This is the real example from the OsirisRex team based on their NAV SPKs. Here is a list of some of the SPKs in alphabetical order:

- orx_190612_190806_190628_od159-N_v1.bsp
- orx_190612_190806_190701_od160-N_v1.bsp
- orx_190612_190806_190703_od161-N_v1.bsp
- orx_190612_190806_190705_od162-N_v1.bsp
- orx_190612_190806_190708_od163-N_v1.bsp
- orx_190612_190806_190719_od167-N_v1.bsp
- orx_190612_190806_190723_od168-N_v1.bsp
- orx_190612_190806_190726_od169-N_v1.bsp
- orx_190612_190813_190730_od170-N_v1.bsp
- orx_190612_190820_190710_od164-N_v1.bsp
- orx_190612_190820_190712_od165-N_v1.bsp
- orx_190612_190820_190716_od166-N_v1.bsp

The SPKs are more complex than most we use currently and contain *both* predicted and reconstructed data. Here's the file naming convention:

**orx_[aaaaaa]_[bbbbbb]_[cccccc]_od[ddd]-N_v[e].bsp**
- **[aaaaaa]** The start date for the reconstructed data as YYMMDD
- **[bbbbbb]** The end date for the predicted data as YYMMDD
- **[cccccc]** The end date for the reconstructed data as YYMMDD. This is also the start date for the predicted data
- **[ddd]** The release version string. This is the version for the entire set of kernels.
- **[e]** The filer version string. This is the version for this specific file and resets to 1 with each kernel set release.

So, for orx_190612_190820_190716_od166-N_v1.bsp we know the following:

- The reconstructed data time range is 2019-06-12 to 2019-07-16
- The predicted data time range is 2019-07-16 to 2019-08-20
- This kernel is from release 166
- This is version 1 of this kernel

After accounting for time ranges and name filtering, we want to use the latest version of the latest release.

#### The problems

There are two problems with how kerneldbgen operates on this example:

1. ISIS does not differentiate between the reconstructed and predicted coverage, it looks at the coverage for *all* of the data in the kernel.
1. Because kerneldb uses reverse-alphabetical order to set priority, the end time for the predicted data will determine which kernels get higher priority within a time range.

For an image taken on 2019-07-17 any of the above kernels could be used because they have some data for that time. As kerneldbgen works right now it will select orx_190612_190820_190716_od166-N_v1.bsp. In actuality, we only want to choose from the following because they have reconstructed coverage for that time:

- orx_190612_190806_190719_od167-N_v1.bsp
- orx_190612_190806_190723_od168-N_v1.bsp
- orx_190612_190806_190726_od169-N_v1.bsp
- orx_190612_190813_190730_od170-N_v1.bsp

In this case, the reverse alphabetical order would select the proper kernel from this subset, orx_190612_190813_190730_od170-N_v1.bsp.

Next let's consider an image taken on 2019-07-04. As with the other image, kerneldbgen just pick the last kernel alphabetically because they all have some coverage for this time. If we filter out based on recon coverage we get the list:

- orx_190612_190806_190705_od162-N_v1.bsp
- orx_190612_190806_190708_od163-N_v1.bsp
- orx_190612_190806_190719_od167-N_v1.bsp
- orx_190612_190806_190723_od168-N_v1.bsp
- orx_190612_190806_190726_od169-N_v1.bsp
- orx_190612_190813_190730_od170-N_v1.bsp
- orx_190612_190820_190710_od164-N_v1.bsp
- orx_190612_190820_190712_od165-N_v1.bsp
- orx_190612_190820_190716_od166-N_v1.bsp

With this filtered list, the incorrect kernel would still be selected. We want to use orx_190612_190813_190730_od170-N_v1.bsp because it is the most recent version of the most recent release that has reconstructed coverage for 2019-07-04. In order to get the correct selection last in the list we would next need to re-order based on release and then version:

- orx_190612_190806_190705_od162-N_v1.bsp
- orx_190612_190806_190708_od163-N_v1.bsp
- orx_190612_190820_190710_od164-N_v1.bsp
- orx_190612_190820_190712_od165-N_v1.bsp
- orx_190612_190820_190716_od166-N_v1.bsp
- orx_190612_190806_190719_od167-N_v1.bsp
- orx_190612_190806_190723_od168-N_v1.bsp
- orx_190612_190806_190726_od169-N_v1.bsp
- orx_190612_190813_190730_od170-N_v1.bsp

## Possible solutions
As there's really 2 problems here we may need a combination of the following

- Add parameters for ordering regexes
- Pass in a file list that gives an explicit ordering priority
- Modify ISIS to differentiate between predicted data coverage and reconstructed data coverage automatically. This may or may not be possible
- Use a config file to control complex behavior
- Provide another customizable tool to do more filtering/re-ordering on the output of kerneldbgen

### PR DESCRIPTION
**Pull request by jessemapel**
Added 3 new arguments to kerneldbgen that can be used instead of (predict/recon/smitthed)dir and (predict/recon/smitthed)filter. These take an explicit list of kernels to use in ascending priority (things later in the list trump things earlier in the list).

## Description

I duplicated the SpiceDbGen::Direct method and then modified it to take a FileList instead of a location and filter. I chose not to split the shared functionality into a separate method because it already uses various helper methods and I'm concerned about readability with another layer of functions.

Writing the lists is a bit complex because they should be written using either `$ISIS3DATA` or better yet `$base` or `$mission` to ensure they are portable between installs. Careful use of sed can help. For example, this would generate the list of MRO SPKs created by the JPL SSD group using the mro110c gravity model:

```
(isis3Ubuntu18.04) jmapel@prog28:/work/users/jmapel/ISIS3/build$ ls $ISIS3DATA/mro/kernels/spk/*mro110c.bsp*
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp10_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp25_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp11_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp26_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp12_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp27_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp13_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp28_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp14_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp29_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp15_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp2_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp16_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp30_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp17_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp31_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp18_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp32_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp19_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp3_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp1_ssd_mro110c.bsp   /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp4_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp20_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp5_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp21_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp6_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp22_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp7_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp23_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp8_ssd_mro110c.bsp
/usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp24_ssd_mro110c.bsp  /usgs/cpkgs/isis3/data/mro/kernels/spk/mro_psp9_ssd_mro110c.bsp
(isis3Ubuntu18.04) jmapel@prog28:/work/users/jmapel/ISIS3/build$ ls $ISIS3DATA/mro/kernels/spk/*mro110c.bsp* | sed "s@$ISIS3DATA/mro@\$mro@g"
$mro/kernels/spk/mro_psp10_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp11_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp12_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp13_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp14_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp15_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp16_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp17_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp18_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp19_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp1_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp20_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp21_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp22_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp23_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp24_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp25_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp26_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp27_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp28_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp29_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp2_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp30_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp31_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp32_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp3_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp4_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp5_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp6_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp7_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp8_ssd_mro110c.bsp
$mro/kernels/spk/mro_psp9_ssd_mro110c.bsp
```

## Related Issue

Fixes #3390.

## Motivation and Context

Kernels are getting more complex and the simple filtering and alpha-numeric sorting used by kerneldbgen right now is not sufficient. See #3390 for a more detailed explanation and examples.

## How Has This Been Tested?

All current tests are passing. A new test has been added for this. It has also been tested with the ORex kernels listed in #3390.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

**Comment by jessemapel** [comment]
Here's a small Python script to fix the end times:

```
import os
import pvl
import re
import datetime

in_file = 'kernels.0001.db'
out_file = 'kernels.0002.db'
kernels = pvl.load(in_file)
for selection in kernels['SpacecraftPointing'].getlist('Selection'):
    res = re.findall("orx_\d{6}_\d{6}_(\d{2})(\d{2})(\d{2})_", selection['File'])
    if not res: continue
    recon_end = datetime.date(2000 + int(res[0][0]), int(res[0][1]), int(res[0][2]))
    selection['Time'][1] = recon_end.strftime('%Y %b %d %H:%M:%S.%f TBD').upper()
pvl.dump(kernels, out_file)
```

It does require the PVL library

File: isis/src/system/apps/kerneldbgen/SpiceDbGen.cpp [hunk 1]
```diff
@@ -131,6 +130,79 @@ PvlObject SpiceDbGen::Direct(QString quality, QString location,
...
+    }
+    else if (grp->name() == type) {
+      grp->setName("Selection");
+      grp++;
+    }
+    else {
+      QString message = "A kernel of type [" + grp->name() + "] has been found in a directory for type [" + type + "]" ;
+      throw IException(IException::Programmer, message, _FILEINFO_);
+      break;
+    }
```
Comment (SgStapleton) [inline isis/src/system/apps/kerneldbgen/SpiceDbGen.cpp:180]: I just want to make sure that I am understanding this if/else block: is this assuming that the first file of valid time range is the type that was intended? Is there ever, for example, a time when the first file is a fluke somehow (with an "SPK" type), but all others have a "CK" type (and the actual type is meant to be "CK")? If so, this error message may be misleading...?

Comment (jessemapel) [reply isis/src/system/apps/kerneldbgen/SpiceDbGen.cpp:180]: Yeah this is some weird logic. It should just check that the name is equal to the type for each group.

**Review by SgStapleton — APPROVED** [review]
This looks good to me.

**Review by krlberry — APPROVED** [review]
:+1: This also looks good to me, and I think it was a good solution to the problem.

**Pull request merged**